### PR TITLE
Support custom tags and variables

### PR DESCRIPTION
**Pull request by guillaume-thisplace**

### CODE REVIEW

File: lib/tags/builtin.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+			'c:set':       (...params) => self.tagSetData.apply(self, params)
+		};
+		this.lastIfMatched = null;
+		this.lastWhenMatched = null;
+	}
+
+	cleanTestText(testExp) {
+		testExp = testExp.trim();
+		if (testExp.startsWith('${') && testExp.endsWith('}')) {
+			testExp = testExp.slice(2, -1)
```
Comment (Comum) [inline lib/tags/builtin.js]: Please add missing semicolon

File: lib/tags/builtin.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+	tagWhenHandler(node, index, data) {
+		let token = node.token
+		let attrText = token.text.slice(5, -1).trim()
+		let attrs = utils.getAttrs(attrText)
+		let testExp = this.cleanTestText(attrs.test);
+		if (!testExp) {
+			throw new Error('no valid test in :when (' + index + ', ' + token.text + ')');
+		}
+		let values = Object.assign({}, this.options.globals, data);
```
Comment (Comum) [inline lib/tags/builtin.js]: Please add missing semicolons in lines 79 - 81 and in 86.

File: lib/tags/builtin.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	tagIfHandler(node, index, data) {
+		let token = node.token
+		let attrText = token.text.slice(5, -1).trim()
+		let attrs = utils.getAttrs(attrText)
+		let testExp = this.cleanTestText(attrs.test);
+		if (!testExp) {
+			throw new Error('no valid test in :if (' + index + ', ' + token.text + ')');
+		}
+		let values = Object.assign({}, this.options.globals, data);
+		let match = utils.evalExp(testExp, values)
```
Comment (Comum) [inline lib/tags/builtin.js]: Please add missing semicolons in lines 107 - 109 and in 115

File: lib/tags/taglib.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	constructor(options, renderer) {
+		this.tagMap = {};
+		this.options = options;
+		this.renderer = renderer;
+	}
+
+	cleanTestText(testExp) {
+		testExp = testExp.trim();
+		if (testExp.startsWith('${') && testExp.endsWith('}')) {
+			testExp = testExp.slice(2, -1)
```
Comment (Comum) [inline lib/tags/taglib.js]: Please add missing semicolon

File: lib/parser.js [hunk 5]
```diff
@@ -1,75 +1,77 @@
...
-					token: token
-				})
-				break
-			case 'directive':
-				node.children.push(parseDirevtive(token))
-				break
-			case 'tag':
-				if (token.closed) {
+		let branch = [];
+		let node = docNode;
```
[the remark this answers is not in the document]
Comment (guillaume-thisplace) [reply lib/parser.js]: OK

File: lib/renderer.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+				return '<!-- removed code: ' + code + ' -->'
+			}
+		}
+	}
+
+	renderCodeExp(node, index, data) {
+		var exp = node.token.data
+		var value = utils.evalExp(exp, this.addGlobals(data));
+		return value !== undefined ? value :
+			node.token.in_tag ? '' : '<!-- removed code_exp: ' + exp + ' --->'
```
[the remark this answers is not in the document]
Comment (guillaume-thisplace) [reply lib/renderer.js]: OK

File: lib/tags/builtin.js (hunk 1 above)
Comment (guillaume-thisplace) [reply lib/tags/builtin.js]: OK

**Pull request merged**